conversion/conversion.go

```diff
@@ -445,26 +440,60 @@ func UpdateDDLForeignKeys(project, instance, dbName string, conv *internal.Conv,
 	}
 	msg := fmt.Sprintf("Updating schema of database %s in instance %s with foreign key constraints ...", dbName, instance)
 	p := internal.NewProgress(int64(len(fkStmts)), msg, internal.Verbose())
-	for i, fkStmt := range fkStmts {
-		// TODO: Improve performance of the foreign key constraint updates.
-		// For example, issue all of the update ops first before waiting for them to complete
-		// so that that can execute in parallel. We could also print out ids of the
-		// long-running operations.
-		op, err := adminClient.UpdateDatabaseDdl(ctx, &adminpb.UpdateDatabaseDdlRequest{
-			Database:   fmt.Sprintf("projects/%s/instances/%s/databases/%s", project, instance, dbName),
-			Statements: []string{fkStmt},
-		})
-		if err != nil {
-			fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
-			conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
-			continue
-		}
-		if err := op.Wait(ctx); err != nil {
-			fmt.Printf("Can't add foreign key with statement %s: %s\n", fkStmt, err)
-			conv.Unexpected(fmt.Sprintf("Can't add foreign key with statement %s: %s", fkStmt, err))
+	type Pair struct {
+		a, b interface{}
+	}
+	fkStmtsQueue := make(chan string, len(fkStmts))
+	results := make(chan Pair, len(fkStmts))
+
+	for _, fkStmt := range fkStmts {
+		fkStmtsQueue <- fkStmt
+	}
+	close(fkStmtsQueue)
+
+	// Spawn `maxWorkers` go routines that each read one fkStatement from the fkStmtsQueue channel.
+	// On succesful creation of a foreign key, push a Pair to the results channel where Pair.b is 1
+	// on success and 0 on failure.
+	for i := int64(1); i <= maxWorkers; i++ {
+		go func(fkStmtsQueue <-chan string, results chan<- Pair) {
+
+			for fkStmt := range fkStmtsQueue {
```

Comment:
this works but a more typical way to implement this is as following:

- define a channel with the same size as the number of workers, also fill the channel with available workers in the beginning; you could call such a channel a pool (as in pool of workers)
- loop over the tasks i.e., fkStmts, and for each task i.e., fkStmt, take out a worker from the pool, and assign the worker the given task and once completed, put the worker back in the pool; to take out a worker from the pool you could use "select from the channel"

Note that worker in this case could be as simple as just integers i.e., pool would be a channel of integers